File: pkce_authentication.py
import json
import jwt
import boto3
import logging
import requests
from jwt import PyJWKClient

# Setup Logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)

# AWS Clients
cognito = boto3.client('cognito-idp')
dynamodb = boto3.client('dynamodb')
s3 = boto3.client('s3')

# Constants (Placeholders for sensitive information)
USER_POOL_ID = "<USER_POOL_ID>"
COGNITO_REGION = "<COGNITO_REGION>"
APP_CLIENT_ID = "<APP_CLIENT_ID>"
COGNITO_TOKEN_URL = f"https://login.example.com/oauth2/token"
COGNITO_JWKS_URL = f"https://cognito-idp.{COGNITO_REGION}.amazonaws.com/{USER_POOL_ID}/.well-known/jwks.json"
DYNAMODB_TABLE_NAME = "<DYNAMODB_TABLE_NAME>"
S3_BUCKET_NAME = "<S3_BUCKET_NAME>"

# Helper Functions
def validate_id_token(token):
    """ Validates JWT ID Token against Cognito JWKs """
    try:
        jwk_client = PyJWKClient(COGNITO_JWKS_URL)
        signing_key = jwk_client.get_signing_key_from_jwt(token).key

        decoded_token = jwt.decode(
            token,
            signing_key,
            algorithms=["RS256"],
            audience=APP_CLIENT_ID,
            options={"verify_aud": True} 
        )

        username = decoded_token.get("cognito:username", decoded_token.get("sub", "unknown_user"))
        return True, decoded_token, username

    except jwt.ExpiredSignatureError:
        return False, {}, "unknown_user"
    except jwt.InvalidTokenError:
        return False, {}, "unknown_user"
    except Exception as e:
        return False, {}, "unknown_user"

def exchange_code_for_tokens(auth_code, code_verifier):
    """ Exchanges Authorization Code for Tokens via Cognito """
    try:
        data = {
            "grant_type": "authorization_code",
            "client_id": APP_CLIENT_ID,
            "code": auth_code,
            "redirect_uri": "http://localhost/redirect.html",
            "code_verifier": code_verifier
        }
        headers = {"Content-Type": "application/x-www-form-urlencoded"}
        response = requests.post(COGNITO_TOKEN_URL, data=data, headers=headers)

        if response.status_code == 200:
            return response.json()
        return None
    except Exception:
        return None

def get_user_cognito_attributes(access_token):
    """ Fetches user attributes from Cognito """
    try:
        response = cognito.get_user(AccessToken=access_token)
        attributes = {attr["Name"]: attr["Value"] for attr in response["UserAttributes"]}
        return attributes, attributes.get("profile", "unknown_user")
    except Exception:
        return {}, "unknown_user"

def get_user_dynamodb_data(username):
    """ Fetches user-related data from DynamoDB """
    try:
        response = dynamodb.get_item(
            TableName=DYNAMODB_TABLE_NAME,
            Key={"TS_user_id": {"S": username}}
        )
        user_data = response.get("Item", {})
        return {k: list(v.values())[0] for k, v in user_data.items()} if user_data else {}
    except Exception:
        return {}

def generate_response(status_code, body, access_token=None, refresh_token=None):
    """ Returns API response with proper headers & cookies """
    if access_token:
        body["access_token"] = access_token
    if refresh_token:
        body["refresh_token"] = refresh_token
    response = {
        "statusCode": status_code,
        "body": json.dumps(body),
        "headers": {
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Credentials": "true",
            "Access-Control-Allow-Methods": "POST, GET, OPTIONS",
            "Access-Control-Allow-Headers": "Content-Type, Authorization"
        }
    }
    if refresh_token:
        response["multiValueHeaders"] = {
            "Set-Cookie": [
                f"refresh_token={refresh_token}; Path=/; Max-Age=2592000; SameSite=None; Secure",
                f"has_refresh_token=true; Path=/; Max-Age=2592000; SameSite=None; Secure"
            ]
        }
    return response

def handle_login(event):
    """ Handles User Login Flow """
    try:
        body = json.loads(event.get("body", "{}"))
        authorization_code = body.get("authorization_code")
        code_verifier = body.get("code_verifier")
        if not authorization_code or not code_verifier:
            return generate_response(400, {"error": "Missing authorization code or verifier"})

        tokens = exchange_code_for_tokens(authorization_code, code_verifier)
        if not tokens:
            return generate_response(401, {"error": "Invalid authorization code"})

        id_token = tokens.get("id_token")
        access_token = tokens.get("access_token")
        refresh_token = tokens.get("refresh_token")

        is_valid, decoded_id_token, username = validate_id_token(id_token)
        if not is_valid:
            return generate_response(401, {"error": "Invalid ID token"})

        email = decoded_id_token.get("email")
        return generate_response(
            200,
            {
                "message": "Login successful",
                "user_id": username,
                "email": email,
                "redirect": "http://localhost/dashboard"
            },
            access_token=access_token,
            refresh_token=refresh_token
        )
    except Exception:
        return generate_response(500, {"error": "Internal Server Error"})

def lambda_handler(event, context):
    """ Main Lambda Handler """
    http_method = event.get("httpMethod", "UNKNOWN")
    if http_method == "OPTIONS":
        return generate_response(200, {})
    if http_method == "POST":
        resource_path = event.get("resource", "Unknown")
        if resource_path == "/auth":
            return handle_login(event)
        return generate_response(404, {"error": "Resource Not Found"})
    return generate_response(405, {"error": "Method Not Allowed"})

File: pkce_authentication.js
document.addEventListener("DOMContentLoaded", async function () {
    // Extract authorization code from URL
    const urlParams = new URLSearchParams(window.location.search);
    const authorizationCode = urlParams.get("code");

    if (!authorizationCode) {
        return;
    }

    // Generate PKCE Code Verifier & Challenge
    const codeVerifier = generateRandomString(128);
    const codeChallenge = await sha256(codeVerifier);

    // Store values for future use
    localStorage.setItem("code_verifier", codeVerifier);
    localStorage.setItem("code_challenge", codeChallenge);

    // API Endpoint (Placeholder for security)
    const authUrl = "<AUTH_API_ENDPOINT>";

    try {
        const requestOptions = {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                "Accept": "application/json",
                "Origin": window.location.origin,
            },
            body: JSON.stringify({
                authorization_code: authorizationCode,
                code_verifier: codeVerifier
            }),
            credentials: "include" 
        };

        const response = await fetch(authUrl, requestOptions);
        const data = await response.json();

        // Store refresh token as a normal cookie
        if (data.refresh_token) {
            document.cookie = `refresh_token=${data.refresh_token}; path=/; max-age=2592000; SameSite=None; Secure`;
            document.cookie = `has_refresh_token=true; path=/; max-age=2592000; SameSite=None; Secure`;
        }

        // Store Access Token in sessionStorage
        if (data.access_token) {
            sessionStorage.setItem("access_token", data.access_token);
        }

        // Store username as a cookie
        if (data.user_id) {
            document.cookie = `username=${data.user_id}; path=/; max-age=2592000; SameSite=None; Secure`;
        }

        // Redirect to dashboard
        setTimeout(() => {
            window.location.href = "<DASHBOARD_URL>";
        }, 3000);
    } catch (error) {
        alert("Authentication failed. Please try again.");
    }
});

// Function to generate a random string (PKCE Code Verifier)
function generateRandomString(length) {
    const possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";
    let randomString = "";
    const randomValues = new Uint8Array(length);
    window.crypto.getRandomValues(randomValues);
    randomValues.forEach(value => {
        randomString += possible.charAt(value % possible.length);
    });
    return randomString;
}

// Function to hash the code verifier for PKCE
async function sha256(plain) {
    const encoder = new TextEncoder();
    const data = encoder.encode(plain);
    const hash = await crypto.subtle.digest('SHA-256', data);
    return btoa(String.fromCharCode(...new Uint8Array(hash)))
        .replace(/\+/g, '-')
        .replace(/\//g, '_')
        .replace(/=+$/, '');
}
